lib/evm/loop.js

```diff
@@ -30,56 +43,45 @@ module.exports = class Loop {
     }
 
     // Initialize internal vm state
-    const runState = await this.initRunState(opts)
+    await this.initRunState(opts)
 
     // Check that the programCounter is in range
-    const pc = runState.programCounter
-    if (pc !== 0 && (pc < 0 || pc >= runState.code.length)) {
+    const pc = this._runState.programCounter
+    if (pc !== 0 && (pc < 0 || pc >= this._runState.code.length)) {
       throw new Error('Internal error: program counter not in range')
     }
 
     let err
     // iterate through the given ops until something breaks or we hit STOP
-    while (this.canContinueExecution(runState)) {
-      const opCode = runState.code[runState.programCounter]
-      runState.opCode = opCode
-      await this.runStepHook(runState)
+    while (this.canContinueExecution()) {
+      const opCode = this._runState.code[this._runState.programCounter]
+      this._runState.opCode = opCode
+      await this.runStepHook()
 
       try {
-        await this.runStep(runState)
+        await this.runStep()
       } catch (e) {
         err = e
         break
       }
     }
 
-    return this.parseVmResults(runState, err, opts.message.gasLimit)
+    return this.parseVmResults(err, opts.message.gasLimit)
   }
 
-  canContinueExecution (runState) {
-    const notAtEnd = runState.programCounter < runState.code.length
-    return !runState.stopped && notAtEnd && !this._result.vmError && !this._result.returnValue
+  canContinueExecution () {
+    const notAtEnd = this._runState.programCounter < this._runState.code.length
+    return !this._runState.stopped && notAtEnd && !this._result.vmError && !this._result.returnValue
   }
 
   async initRunState (opts) {
-    const runState = {
+    Object.assign(this._runState, {
       code: opts.message.code,
-      stopped: false,
-      programCounter: opts.pc | 0,
-      opCode: undefined,
-      opName: undefined,
+      programCounter: opts.pc | this._runState.programCounter,
       gasLeft: new BN(opts.message.gasLimit),
-      memory: new Memory(),
-      memoryWordCount: new BN(0),
-      stack: new Stack(),
       validJumps: this._getValidJumpDests(opts.message.code),
-      highestMemCost: new BN(0),
-      lastReturned: [],
-      // TODO: Replace with EEI methods
-      _common: this._vm._common,
-      stateManager: this._state._wrapped,
-      storageReader: opts.storageReader || new StorageReader(this._state._wrapped)
-    }
+      storageReader: opts.storageReader || this._runState.storageReader
+    })
```

Comment:
Nice to have this base value initialization and runtime value assignment separation.